### PR TITLE
APS-1014 Back link handler

### PR DESCRIPTION
**Pull request by bobmeredith**
# Context

https://dsdmoj.atlassian.net/browse/APS-1014

# Changes in this PR

Adds a utility to supply back-links for problematic pages.
The fn takes the request, a list of acceptable parent paths and the path of the page. If the request referrer matches one of the parent paths, the referrer is returned and stored in the session. If the referrer doesn't match a parent path, the last stored parent referrer is returned instead.
This means. that a page that can be called from multiple points can always have a back-link to the page it last came from even if there are page transitions on that page e.g. if the user changes tab, alters filters or follows an 'action' journey.  

This PR adds the above functionality to:

- Placement view page
- Out of service bed listing page
- Out of service bed details page
- New withdrawal page 

## Screenshots of UI changes

No UI change

### CODE REVIEW

**Review by froddd — APPROVED** [review]
Seems to work well in practice, there may be a couple of bits that can be cleaned up further (using `request` to provide current URL pattern), and possibly an edge case to consider, but this is a huge improvement as it is!

File: server/services/sessionService.test.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    it('should return the referer if it matches a provided path', () => {
+      const request = mockRequest(matchingReferer, undefined)
+      expect(service.getPageBackLink(pagePattern, request, matchList)).toEqual(matchingReferer)
+      expect(request.session.pageReferers[pagePattern]).toEqual(matchingReferer)
+    })
+
+    it('should return the stored referer if the current referer does not match a path', () => {
+      const request = mockRequest(nonMatchingReferer, lastStoredReferer)
+      expect(service.getPageBackLink(pagePattern, request, matchList)).toEqual(lastStoredReferer)
+    })
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/services/sessionService.test.ts:35]: Yes, that's a good question.
The match patterns don't contain any ids so they can't in general be used out of context.
I guess what we could so is:

- Provide a default where possible i.e. where there are no ids in the parent page.
- Remove the backlink from the UI and let users find their way back - perhaps through their browser history.
- Link to the homepage.

File: server/services/sessionService.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+   * Provides referer links for pages with multiple routes
+   * A set of match path patterns are provided. If the referer in the request matches one of these paths,
+   * the referer is returned and stored in the session against this page pattern.
+   * If no match, the last stored referer for this page is returned
+   * @param pagePattern the pattern of the path of the page being rendered.
+   * @param request the page request object
+   * @param matchList an array of path.pattern strings to match against the referer in the request
+   * @return the url to use as the backlink for the page.
+   */
+  getPageBackLink = (pagePattern: string, req: Request, matchList: Array<Path>): string => {
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/services/sessionService.ts:17]: ooh nice one - That will tidy things up a bit!

File: server/services/sessionService.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    const {
+      session,
+      headers: { referer },
+    } = req
+    const refererPath = (referer && new URL(referer).pathname) || ''
+    const foundReferer = matchList.find(path => match(path)(refererPath))
+    const lastReferer = session?.pageReferers?.[pagePattern]
+    if (session && foundReferer && lastReferer !== referer) {
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/services/sessionService.ts:25]: I guess not in reality. I probably put that in while fixing tests where it wasn't fully mocked. Will remove.

File: server/controllers/manage/outOfServiceBedsController.test.ts [hunk 4]
```diff
@@ -25,28 +25,30 @@ import {
...
+  const backLink = '/back/link'
   let request: DeepMocked<Request>
   const response: DeepMocked<Response> = createMock<Response>({})
   const next: DeepMocked<NextFunction> = createMock<NextFunction>({})
 
   const outOfServiceBedService = createMock<OutOfServiceBedService>({})
   const premisesService = createMock<PremisesService>({})
   const apAreaService = createMock<ApAreaService>({})
+  const sessionService = createMock<SessionService>()
+  sessionService.getPageBackLink.mockReturnValue('back/link/')
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/controllers/manage/outOfServiceBedsController.test.ts:45]: Fair enough

File: server/services/sessionService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/services/sessionService.ts:17]: Ah, that doesn't work!
Actually, the page path is just a unique identifier for the page. It doesn't have to be the path - but I thought that would be a convenient id. 
Some of our pages though have the same controller routed from different paths - notably the tabs on the placement detail page. That breaks things badly!
 I _could_ make the path an optional parameter for cases like that but it's probably safest that a literal is passed from each controller.
I'll update the docs on `getPageBackLink` to make that clear.

[the remark this answers is not in the document]
Comment (froddd) [reply server/services/sessionService.ts:17]: If it's used as a unique id for the page (for storing it against the session) then maybe the actual path should be used, rather than the pattern? Potentially, during a single session, a user could visit 2 pages with the same pattern but coming from different referers -- this would resolve both pages to the same referer, is that correct?

[the remark this answers is not in the document]
Comment (froddd) [reply server/services/sessionService.ts:17]: If it's only used as a unique ID, then maybe it should be the actual path? This would be a definite edge case, but during the same session could a user visit 2 different pages with a similar pattern, but from different referers -- this would effectively overwrite one or the other?

[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/services/sessionService.ts:17]: Yes, it would overwrite, but I think that's fine.

My logic is that the user can only get to a specific page (page controller) from one of the designated parents. They may make many page transitions after that - from and back to the page in question, but they would still want the back-button on the page to point to this last visited designated parent.
If they navigate such that they hit the page from a different designated parent, we would want the back button to switch to the new parent so overwriting in the session is what we want.

So the unique Id is to identify the page controller rather than a specific path. We can't use the actual page path because that doesn't map to the page controller. Neither sadly does the route pattern because of the cases where we have multiple routes pointing to the same controller - the big offender here being the placement view with its tabs.

[the remark this answers is not in the document]
Comment (froddd) [reply server/services/sessionService.ts:17]: Thanks for clarifying, this makes sense!
